File: 859.buddy-strings.cpp
/*
 * @lc app=leetcode id=859 lang=cpp
 *
 * [859] Buddy Strings
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  bool buddyStrings(string a, string b) {
    int alen = a.length();
    int blen = b.length();
    if (alen != blen) {
      return false;
    }
    if (a == b) {
        int size = std::set<char>(a.begin(), a.end()).size();
        if (size != alen) {
            return true;
        }
        return false;
    }
    int firsta = -1;
    int firstb = -1;
    int seconda = -1;
    for (int i = 0; i < alen; i++) {
      if (a[i] != b[i]) {
        if (firsta < 0) {
          firsta = a[i];
          firstb = b[i];
        } else if (seconda < 0) {
          if (a[i] != firstb || b[i] != firsta) {
            return false;
          }
          seconda = a[i];
        } else {
          return false;
        }
      }
    }
    if (seconda < 0) {
      return false;
    }
    return true;
  }
};
// @lc code=end

int main() {
  Solution s;
  ASSERT_EQ(s.buddyStrings("ab", "a"), false);
  ASSERT_EQ(s.buddyStrings("ab", "ba"), true);
  ASSERT_EQ(s.buddyStrings("ab", "ab"), false);
  ASSERT_EQ(s.buddyStrings("aa", "aa"), true);
  ASSERT_EQ(s.buddyStrings("aaaaaaabc", "aaaaaaacb"), true);
  return 0;
}

File: 859.buddy-strings.rs
/*
 * @lc app=leetcode id=859 lang=rust
 *
 * [859] Buddy Strings
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn buddy_strings(s: String, goal: String) -> bool {
        if s.len() != goal.len() {
            return false;
        }
        if s == goal {
            return s
                .chars()
                .fold(
                    std::collections::HashMap::<char, usize>::new(),
                    |mut m, count| {
                        *m.entry(count).or_default() += 1;
                        m
                    },
                )
                .into_iter()
                .any(|(_, v)| v > 1);
        }

        let mut state = 0;
        let mut last_pair = (0, 0);
        s.as_bytes()
            .iter()
            .zip(goal.as_bytes())
            .filter(|(&sc, &gc)| sc != gc)
            .try_for_each(|(&sc, &gc)| match state {
                0 => {
                    last_pair = (sc, gc);
                    state = 1;
                    Ok(())
                }
                1 if last_pair.0 == gc && last_pair.1 == sc => {
                    state = 2;
                    Ok(())
                }
                _ => {
                    state = 0;
                    Err(())
                }
            });
        state == 2
    }
}
// @lc code=end

#[test]
fn test1() {
    assert!(Solution::buddy_strings("ab".to_string(), "ba".to_string()));
}

#[test]
fn test2() {
    assert!(!Solution::buddy_strings("ab".to_string(), "ab".to_string()));
}

#[test]
fn test3() {
    assert!(Solution::buddy_strings("aa".to_string(), "aa".to_string()));
}

#[test]
fn test4() {
    assert!(!Solution::buddy_strings(
        "abcd".to_string(),
        "badc".to_string()
    ));
}
